### DISCUSSION TITLE
Error with malloc and memory limit

### DISCUSSION THREAD

**Discussion by serman** [climateinteractive/SDEverywhere#522] [Q&A] [answered]
HelloI 
I am posting here since it could be useful for others too. 
_(edit: sorry,  after submitting, I am realising this message should go to the discussion board)_

have received a new version of the model and I found a "random" error related with the outputs.
The model is generated correctly but when I run it I find this error:
 
![image](https://github.com/user-attachments/assets/ef7fd021-8bc6-4620-98f0-59f740631431)

Edited by Chris: Adding a text snippet of the error message here for searchability:
```
RuntimeError: Aborted(OOM). Build with -s ASSERTIONS=1 for more info.
  (stack trace omitted)
```

After removing some variables form the output.csv file I realized that the problem was not caused by any of them but by having too many. I managed to add the -S ASSERTIONS=1 to the wasm plugin:  

```javascript
wasmPlugin({ 
emccArgs: [
"-Wall", "-Os ", 
"-sSTRICT=1 ",
 "-sMALLOC=emmalloc", 
"-sFILESYSTEM=0 ",
 "-sMODULARIZE=1 ",
 "-sSINGLE_FILE=1 ", 
"-sEXPORT_ES6=1 ",
 "-sUSE_ES6_IMPORT_META=0 ", 
"-sENVIRONMENT='web,webview,worker' ", 
"-sEXPORTED_FUNCTIONS='_malloc','_free','_getInitialTime','_getFinalTime','_getSaveper','_setLookup','_runModelWithBuffers']", 
"-sEXPORTED_RUNTIME_METHODS=['cwrap']",
 "-sASSERTIONS=1"] }),
```

_(note: I recommend including an example of this in the documentation since I didn't know that -s params don't have a space  between "-s XXXXPARAMXXX" I can add it if you want)_

And the error I  got was: 
<img width="679" alt="image" src="https://github.com/user-attachments/assets/68ce91b5-26f3-4c88-a614-06838dca71c6">

Edited by Chris: Adding a text snippet of the error message here for searchability:
```
RuntimeError: Aborted(Cannot enlarge memory arrays to size 19877536 bytes (00M). Either
(1) compile with -s INITIAL_MEMORY=X with X higher than the current value 16777216,
(2) compile with -s ALLOW_MEMORY_GROWTH=1 which allows increasing the size at runtime, or
(3) if you want malloc to return NULL (0) instead of this abort, compile with -s ABORTING_MALLOC=0 )
```

Which, as suspected, is due to excess memory consumption.
Adding: `-sINITIAL_MEMORY=32MB`,or  `-sALLOW_MEMORY_GROWTH=1`

seems to fix the problem. 
What do you recommend? Do you think we can get into trouble by using 32MB? It doesn't seem like a lot for an application but I am not familiar with WASM limitations.
Rgds

**Comment by chrispcampbell** [comment, marked as answer]
Hi @serman, thanks for submitting this.  I agree that this info can be helpful to others who might run into similar issues, so I converted this to a discussion thread and edited your post to include text versions of the error messages you saw for better searchability.

For the `-s` args, I filed a separate issue #523 to improve the documentation.

I think that either solution, `-sINITIAL_MEMORY=32MB` or `-sALLOW_MEMORY_GROWTH=1` would be fine.  I don't think 32MB should be too much for modern browsers/devices.  We haven't hit the limit with the default settings in En-ROADS or C-ROADS, but I'm not surprised for it to happen with large models like yours.

We could consider changing the default args to include `-sALLOW_MEMORY_GROWTH=1`, which is most flexible, but might make it harder to detect memory leaks.  (Personally I prefer to have `ALLOW_MEMORY_GROWTH` disabled for that reason.)

I will think about this some more and will consider filing an issue about changing the default parameters, but for now, glad that you were able to figure out emcc args that work well for your use case.

**Comment by serman** [comment]
Thanks for the editing. 

I think the main problem is not the model size (which is -of course- important) but the fact that I need to read many variables for 8 regions. (subscripts) Currently I have 491 variables in the outputs.csv

I also noticed that in the latest model version that I received, the "SAVEPER" variable was incredibly small. So It was trying to allocate memory space for too many points for each variable.

After I reduced the SAVEPER to 1 ( I don't need more than one datapoint per year) It seems It doesn't reach the memory limit.
```
SAVEPER  = 1
	~	Year [0,?]
	~	The frequency with which output is stored.
	|
``
